maas/resource_maas_vlan_dhcp.go

```diff
@@ -0,0 +1,265 @@
+		ReadContext:   resourceVLANDHCPRead,
+		UpdateContext: resourceVLANDHCPUpdate,
+		DeleteContext: resourceVLANDHCPDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: func(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Schema: map[string]*schema.Schema{
+			"fabric": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: "Database ID of the fabric of the VLAN whose DHCP is managed. This parameter `fabric` and `vlan` are used to identify the VLAN.",
+			},
+			"ip_ranges": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "A set of IP range ids to server DHCP to. IP ranges must be of type dynamic.",
+				Elem: &schema.Schema{
+					Type: schema.TypeInt,
+				},
+			},
+			"primary_rack_controller": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"relay_vlan"},
+				AtLeastOneOf:  []string{"primary_rack_controller", "relay_vlan"},
+				Description:   "The system_id of the Rack controller to to use as primary for DHCP.",
+			},
+			"relay_vlan": {
+				Type:          schema.TypeInt,
+				Optional:      true,
+				ConflictsWith: []string{"primary_rack_controller", "secondary_rack_controller"},
+				AtLeastOneOf:  []string{"primary_rack_controller", "relay_vlan"},
+				Description:   "Database ID of the VLAN to to use as a relay for DHCP.",
+			},
+			"secondary_rack_controller": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				RequiredWith:  []string{"primary_rack_controller"},
+				ConflictsWith: []string{"relay_vlan"},
+				Description:   "The system_id of the Rack controller to to use as secondary for DHCP.",
+			},
+			"subnets": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "A set of subnet ids to serve DHCP on their dynamic IP ranges.",
+				Elem: &schema.Schema{
+					Type: schema.TypeInt,
+				},
+			},
+			"vlan": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: "VID of the VLAN whose DHCP is managed. This parameter `vlan` and `fabric` are used to identify the VLAN.",
+			},
+		},
+	}
+}
+
+func resourceVLANDHCPCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+	// Validation
+	err := confirmAllIPRangesDynamic(client, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = confirmAllSubnetsWithADynamicIPRange(client, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = confirmIPRangeSubnetsInVLAN(client, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	// Turn on DHCP
+	fabricID := d.Get("fabric").(int)
+	vlanID := d.Get("vlan").(int)
+	params := getVLANDHCPParams(d)
+
+	_, err = client.VLAN.Update(fabricID, vlanID, params)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(strconv.Itoa(vlanID))
```

Comment:
vlanID here is the vid and not the database id. As such, it is not unique because you can have the same vid for many vlans. Most probably you have to set the vlan database id

Reply:
Or, to align better with API v2 endpoints which require fabric+vid to uniquely identify the vlan, you could use their combination with a separator for the resource ID. This one can also help with the acceptance tests where we get this ID and perform the validation logic.